src/core/thread/mesh_forwarder.cpp

```diff
@@ -206,6 +214,173 @@ void MeshForwarder::ScheduleTransmissionTask()
+        {
+            mMac.EnableSrcMatch(false);
+            mSrcMatchEnabled = false;
+        }
+        else
+        {
+            // succeed in adding to source match table
+            aChild.mAddSrcMatchEntryPending = false;
+        }
+    }
+
+exit:
+
+    if (!mSrcMatchEnabled)
+    {
+        if (aChild.mAddSrcMatchEntryPending && (aChild.mQueuedIndirectMessageCnt == 1))
+        {
+            if (aChild.mAddSrcMatchEntryShort)
+            {
+                mSrcMatchOverflowShortCnt++;
+            }
+            else
+            {
+                mSrcMatchOverflowExtCnt++;
+            }
+
+        }
+
+        otLogDebgMac("Overflow(Short,Ext)=(%d, %d), Released(Short, Ext)=(%d,%d)\n",
+                     mSrcMatchOverflowShortCnt,
+                     mSrcMatchOverflowExtCnt,
+                     mSrcMatchReleasedShortCnt,
+                     mSrcMatchReleasedExtCnt);
+    }
+
+    return;
+}
+
+void MeshForwarder::ClearSrcMatchEntry(Child &aChild)
+{
+    Mac::Address macAddr;
+
+    if (aChild.mAddSrcMatchEntryShort)
+    {
+        macAddr.mLength = sizeof(macAddr.mShortAddress);
+        macAddr.mShortAddress = aChild.mValid.mRloc16;
+    }
+    else
+    {
+        macAddr.mLength = sizeof(macAddr.mExtAddress);
+        memcpy(macAddr.mExtAddress.m8, aChild.mMacAddr.m8, sizeof(macAddr.mExtAddress));
+    }
+
+    if (mMac.ClearSrcMatchEntry(macAddr) == kThreadError_None)
+    {
+        // if waiting for more source table resources
+        if (!mSrcMatchEnabled)
+        {
+            if (aChild.mAddSrcMatchEntryShort)
+            {
+                mSrcMatchReleasedShortCnt++;
+            }
+            else
+            {
+                mSrcMatchReleasedExtCnt++;
+            }
+        }
+    }
+    else
+    {
+        // if finished queued messages for SED which is not added into the source match table
+        if (aChild.mAddSrcMatchEntryShort)
+        {
+            mSrcMatchOverflowShortCnt--;
+        }
+        else
+        {
+            mSrcMatchOverflowExtCnt--;
+        }
+
+        aChild.mAddSrcMatchEntryPending = false;
+    }
+
+    otLogDebgMac("Overflow(Short,Ext)=(%d, %d), Released(Short, Ext)=(%d,%d)\n",
+                 mSrcMatchOverflowShortCnt,
+                 mSrcMatchOverflowExtCnt,
+                 mSrcMatchReleasedShortCnt,
+                 mSrcMatchReleasedExtCnt);
+
+    // use tasklet for re-enable source match to avoid possible disorder due to
+    // possible repeatedly disable/enable when many SEDs exists
+    mUpdateSrcMatchTask.Post();
+}
+
+void MeshForwarder::HandleUpdateSrcMatchTask(void *aContext)
+{
+    MeshForwarder *obj = reinterpret_cast<MeshForwarder *>(aContext);
+    obj->HandleUpdateSrcMatchTask();
+}
+
+void MeshForwarder::HandleUpdateSrcMatchTask()
+{
+    // Re-enable source match if available source match table resources are enough
+    if (!mSrcMatchEnabled &&
+        ((mMac.IsSrcMatchTableShared() &&
```

Comment:
I'm not sure this properly captures all radio implementations.  For example, while the CC2538 may be able to carry two shorts for every extended, other implementations may be able to carry four shorts for every extended.  I know some radios will hash the extended down to 2 bytes, meaning that the overhead for each short and extended entry is the same.

I think we should consider the following approach:  Whenever we release an entry in the overflow state, first try to add as many pending shorts as we can to the table then, if there is still space available, try to add as many extended addresses to the table.  If after the adds there are no more pending addresses, then we are no longer overflowing and can enable source match again.